sbt: Scala version handling

Locke opened the pull request:
There are currently some workarounds in the sbt settings to exclude some modules from building with Scala 2.11, mostly for Jetty12 (6dc1987f, 912527f2).

We already have the `lociProject.scala2only` helper to define modules only for Scala 2.x.
When cross building the full matrix, building with Scala 3 already skips the compile and publish task (note: [1]). This PR adds also skipping the test task.

This PR adds a similar `scala 2.12+` helper, to define a module requires at least Scala 2.12.
For that, the builder is generalized, to use a generic Scala version check.

Instead of implicitly skipping compile & publish from a setting that is pulled in from a dependency definition, this will more explicitly state that a module is only for Scala 2.12+.

~~This also allows us to depend on lociCommunicatorWsJetty12 for the aggregation / publish.~~
Jetty 12 tests are still flaky. CI can be restarted for the failed jobs, but it took me 5 attempts to have them all pass. So I disabled the Jetty 12 tests again.

[1] sbt-crossproject does not support varying Scala versions per module - skipping some tasks is a workaround. sbt-projectmatrix supports varying Scala versions per module, but it may need some more work to migrate to that plugin. I started some work with the first two commits, which should not have any effect to the current usage with sbt-crossproject.


Creating as WIP/Draft, as I have not yet tested this with the full matrix.
There is at least one test failing currently, which I need to investigate: https://github.com/Locke/scala-loci/actions/runs/8250171160/job/22564251663

pweisenburger commented:
This is very cool :-)

c05140a
> This leads to `-Xlint` being present when switching the Scala version via `++3.3`.
> Note: switching the Scala version via `++3.3` is still not supported, but this is a step towards it.

Good catch! What is the remaining problem with `+3.3.`?

Locke commented:
I rebased this on top of your `only` addition and updated / reordered the commits slightly.

> [...] What is the remaining problem with ++3.3.?

I think I wrote that part of the commit message, when I had a different `crossScalaVersions` for some projects (instead of compile / skip like it is now) and did not yet know that this is not supported by sbt-crossproject. At that state, sbt mixed 2.12 classpaths with 3.3 which had some strange errors.

I tested this again today with and without your `only` addition and did not have that issue.


I swapped the first two commits to make the reason for b9473d5f99 more clear (e.g., if you checkout just the first commit and look at `show scalacOptions`) and updated the commit message. Anyhow, these two commits are totally optional for now and just a preparation in case 
you want to switch from sbt-crossproject to sbt-projectmatrix.


I'm removing the WIP flag, as I added that for failing tests that can now be restarted more easily with my previous already merged PR.

pweisenburger commented:
Thanks for the work! I merged the changes except the first two commits. Let's revisit these changes when/if sbt merges sbt-projectmatrix and stick to "compile / skip" for now.
